torch/csrc/utils/tensor_flatten.cpp

```diff
@@ -1,44 +1,93 @@
 #include "torch/csrc/utils/tensor_flatten.h"
 
+#include <map>
 #include <unordered_map>
 
-namespace torch { namespace utils {
+namespace torch {
+namespace utils {
 
 using namespace at;
 
 std::vector<TensorGroup> take_tensors(TensorList tensors, size_t size_limit) {
   std::vector<TensorGroup> results;
-  results.reserve(tensors.size()); // an overapproximation, but at least we won't have to copy stuff around
-  std::unordered_map<at::Type*, TensorGroup> groups;
-  for (const auto & tensor : tensors) {
-    auto & type = tensor.type();
+  // an overapproximation, but at least we won't have to copy stuff around
+  results.reserve(tensors.size());
+  std::map<TypeID, TensorGroup> groups;
+  for (const auto& tensor : tensors) {
+    auto& type = tensor.type();
     size_t tensor_size;
     if (type.is_sparse()) {
       const auto& indices = tensor._indices();
       const auto& values = tensor._values();
       tensor_size = indices.numel() * indices.type().elementSizeInBytes() +
-                    values.numel() * indices.type().elementSizeInBytes();
+          values.numel() * indices.type().elementSizeInBytes();
     } else {
       tensor_size = tensor.numel() * type.elementSizeInBytes();
     }
-    auto & type_group = groups[&type];
+
+    auto& type_group = groups[type.ID()];
     type_group.tensors.push_back(tensor);
     type_group.size += tensor_size;
-    if (type_group.size + tensor_size >= size_limit) {
+    if (type_group.size >= size_limit) {
       results.emplace_back();
       std::swap(results.back(), type_group);
     }
   }
   // End case. Look for any remaining groups and return them.
-  for (auto & entry : groups) {
-    auto & group = entry.second;
+  for (auto& entry : groups) {
+    auto& group = entry.second;
     if (group.size > 0) {
       results.emplace_back(std::move(group));
     }
   }
   return results;
 }
 
+std::vector<TensorGroup> take_tensors_finegrained(
+    TensorList tensors,
```

Comment:
How does this differ from the function above? There’s a lot of code duplication

Reply:
Eventually, I would like to use the function for BW bucketing strategy, the difference is explained in comments, I will rewrite the comments to make it more clear